app/s3_client/s3_letter_upload_client.py

```
@@ -81,7 +81,9 @@ def get_letter_s3_object(service_id, file_id):
         return s3.Object(current_app.config['TRANSIENT_UPLOADED_LETTERS'], file_location).get()
     except botocore.exceptions.ClientError as e:
         if e.response['Error']['Code'] == 'NoSuchKey':
-            raise LetterNotFoundError(f'Letter not found for service {service_id} and file {file_id}')
+            message = f'Letter not found for service {service_id} and file {file_id}'
+            current_app.logger.warning(message)
```

Loosely held opinion, but in general i prefer to have exception logs higher up - ie: we should only log the exception when we're actively suppressing the exception's bubble-up. That way, the function calling can decide whether it's important if the exception is thrown or not. It also means that when calling code, we don't need to worry "are any potential exceptions already logged?" - as either we will write the handling code ourselves or let it bubble up further to app level handlers where things will get logged.

For example, hypothetically we might decide it's a warning if `uploaded_letter_preview` fails but an error if `send_uploaded_letter`.

so in summary, i'd rather we kept this file as it is and kept the log lines in uploads.py, but just downgraded them to `warning` instead of `error`.

Thoughts?

---

That's a fair point. I was attempting to DRY-up the log but it doesn't save much and I agree with your logic - logic I normally follow myself but for wanting to de-duplicate this bit of code. On reflection, it's the whole block of code that would benefit from a refactor, but that's out-of-scope (and it's only repeated once so arguably not an issue).

I've amended the commit to simply downgrade the second instance of the log.